### ISSUE TITLE
Shutdown method on server do not close the server

### ISSUE THREAD

**Issue by Crevil** [go-swagger/go-swagger#1316]
## Problem statement
When generating a server implementation and calling `Server.Shutdown` method from `restapi` package the server is not actually closed though the documentation states `Shutdown server and clean up resources`. Only the `API.ServerShutdown` callback from `operations` package is called but the actual server is not closed.

The graceful server exposes the API needed to shut it down gracefully but the generated abstraction does not allow access to this API. Because of this, only an OS signal can kill the graceful server when running the generated server.

I propose to expand the `Server.Shutdown` method to signal the graceful server to initiate a shutdown. It is a fairly simple addition to the API and I'm happy to make a PR with the changes.

When looking over the source the following changes are necessary:
- `Server.Serve` should listen for on a shutdown channel and stop the graceful server when receiving a signal
- `Server.Shutdown` should send a signal on the shutdown channel
- Code on examples should be updated to hold the new implementation

## Steps to reproduce
A slightly modified version of the [`custom-server` tutorial example](https://github.com/go-swagger/go-swagger/tree/master/examples/tutorials/custom-server) shows the behaviour. After 5 seconds the server is instructed to shut down by calling `Shutdown`. The `ServerShutdown` callback is called but the server continuous to run. 

When running the output is as follows. Not that I manually interrupt the server with CTRL + C.
```
$ go run ./cmd/greeter/main.go --port 3000
2017/12/08 14:03:39 Serving greeter at http://[::]:3000
2017/12/08 14:03:44 api shutdown called
^C2017/12/08 14:03:53 shutdown initiated
2017/12/08 14:03:53 Stopped serving greeter at http://[::]:3000
```

Below is a stripped snippet from the `main.go`.

```go
// from  examples/tutorials/custom-server/cmd/greeter/main.go

// create new service API
api := operations.NewGreeterAPI(swaggerSpec)
api.ServerShutdown = func() { log.Println("api shutdown called") } // added
server := restapi.NewServer(api)

// parse flags
flag.Parse()
// set the port this service will be run on
server.Port = *portFlag

// GetGreetingHandler greets the given name,
// in case the name is not given, it will default to World
api.GetGreetingHandler = operations.GetGreetingHandlerFunc(
	func(params operations.GetGreetingParams) middleware.Responder {
		// ...
	})

// added
go func() {
	time.Sleep(5 * time.Second)
	server.Shutdown()
}()

// serve API
```

**Comment by Crevil** [comment]
@casualjim Would you be interested in this PR?
I belive this is more of a bug fix than an enhancement as the generated documentation describes that the server should shutdown after a call to `Server.Shutdown()` but it does not?

**Comment by fredbi** [comment]
My own feeling is that all contributions are welcome.
You should not worry that much about bug vs enhancement distinction in issues.
It is more an indication on how to priorize things.
Since the toolkit has many use cases it is often difficult to judge priority from a single userr's perspective.

Generally speaking if you feel it is important, post a PR.

**Comment by Crevil** [comment]
The reason I'm asking is to prioritise my time as well. We are using this in production now and would like to know if a PR is likely to be released any time soon.

Any how, I'll prepare a PR. Any checklists or alike I should use? I had some difficulty running the different make targets locally. Is it fine to just make the changes in the templates and sources and let example generation etc. be handled when things are to be merged?

**Comment by casualjim** [comment]
yes I think this is a valuable contribution. I have to see exactly where the changes will be made and at worst I'd ask to rename a method or something.

**Comment by casualjim** [comment]
if you can run `go test ./...` we should be good

**Comment by Crevil** [comment]
Great. I'll prepare the PR. Thank you for the quick answers.
